pimd/pim_register.c

```diff
@@ -131,32 +156,58 @@ int pim_register_stop_recv(struct interface *ifp, uint8_t *buf, int buf_size)
 	pim_parse_addr_ucast(&source, buf, buf_size);
 	sg.src = source.u.prefix4;
 
+	rp = RP(pim_ifp->pim, sg.grp);
+
+	if (sg.src.s_addr == rp->rpf_addr.u.prefix4.s_addr)
+		sg.src.s_addr = INADDR_ANY;
+
 	upstream = pim_upstream_find(pim, &sg);
-	if (!upstream) {
+
+	if ((!upstream) && (sg.src.s_addr != INADDR_ANY))
 		return 0;
-	}
 
-	if (PIM_DEBUG_PIM_REG)
-		zlog_debug("Received Register stop for %s", upstream->sg_str);
+	/*
+	 * RFC 7761 Sec 4.4.1
+	 * Handling Register-Stop(*,G) Messages at the DR:
+	 *   A Register-Stop(*,G) should be treated as a Register-Stop(S,G) for
+	 * all (S,G) Register state machines that are not in the NoInfo state.
+	 */
+
+	/* No starg upstream. loop upstream list to get all sg */
+	if ((!upstream) && (sg.src.s_addr == INADDR_ANY)) {
+		struct pim_upstream *up;
+
+		frr_each (rb_pim_upstream, &pim->upstream_head, up) {
+			if (sg.grp.s_addr != up->sg.grp.s_addr)
+				continue;
+
+			if (PIM_DEBUG_PIM_REG)
+				zlog_debug("Executing Reg stop for %s",
+					   up->sg_str);
+
+			pim_reg_stop_upstream(pim, up);
+		}
 
-	switch (upstream->reg_state) {
-	case PIM_REG_NOINFO:
-	case PIM_REG_PRUNE:
 		return 0;
-	case PIM_REG_JOIN:
-		upstream->reg_state = PIM_REG_PRUNE;
-		pim_channel_del_oif(upstream->channel_oil, pim->regiface,
-				    PIM_OIF_FLAG_PROTO_PIM, __func__);
-		pim_upstream_start_register_stop_timer(upstream, 0);
-		pim_vxlan_update_sg_reg_state(pim, upstream,
-			false/*reg_join*/);
-		break;
-	case PIM_REG_JOIN_PENDING:
-		upstream->reg_state = PIM_REG_PRUNE;
-		pim_upstream_start_register_stop_timer(upstream, 0);
+	}
+
+	/* Starg upstream present. Loop through children */
+	if (upstream && (sg.src.s_addr == INADDR_ANY)) {
```

Comment:
the sg.src.s_addr test is superfluous just do a loop over all list elements in upstream->sources.  If it's a s,g it will be an empty list.

Reply:
This makes good readability as well as using that library macro to figure out starg or sg will take additional instruction.